File: combinatorics/resource/TMD/graph_tree/Code/problem11/problem11.py
from typing import Dict, List, Optional, Set, Tuple

class DfsSolverPython:
    """
    Encapsulates both recursive and iterative DFS implementations in Python.

    This class provides a structured way to run DFS on a graph. It holds the graph
    and the results of the traversal. Each call to a `solve_` method resets the
    state, allowing for multiple runs on the same graph instance.
    """
    
    def __init__(self, graph: Dict[int, List[int]]):
        """
        Initializes the solver with a graph.

        :param graph: Adjacency list representation of the simple graph.
                      Example: {0: [1, 2], 1: [0, 3], ...}
        """
        self.graph = graph
        # Determine all unique nodes present in the graph
        self.nodes = set(graph.keys())
        for neighbors in graph.values():
            for neighbor in neighbors:
                self.nodes.add(neighbor)
        
        # These will be reset for each run
        self.parent: Dict[int, Optional[int]] = {}
        self.visited: Set[int] = set()

    def solve_recursive(self, start_node: int) -> Tuple[Dict[int, Optional[int]], List[int]]:
        """
        Public method to start the recursive Depth-First Search.

        This function sets up the necessary data structures (parent map, visited set)
        and then calls the private utility function to perform the actual recursion.
        
        :param start_node: The node to start the search from.
        :return: A tuple containing the parent map for path reconstruction and
                 the list of nodes in the order they were visited.
        """
        self.parent = {node: None for node in self.nodes}
        self.visited = set()
        traversal_path = []

        if start_node in self.nodes:
            self._dfs_recursive_util(start_node, traversal_path)
            
        return self.parent, traversal_path

    def _dfs_recursive_util(self, u: int, path: List[int]):
        """
        Recursive helper function that embodies the DFS logic.

        This is the direct implementation of the recursive formula:
        1. Mark and process the current node `u`.
        2. For each unvisited neighbor `v`, set its parent and recurse on `v`.
        The system's call stack handles the LIFO (Last-In, First-Out) behavior.
        
        :param u: The current node to visit.
        :param path: The list to record the traversal order.
        """
        # Mark the current node as visited and add to the traversal path.
        self.visited.add(u)
        path.append(u)

        # Recur for all unvisited vertices adjacent to this vertex.
        for v in self.graph.get(u, []):
            if v not in self.visited:
                self.parent[v] = u
                self._dfs_recursive_util(v, path)

    def solve_iterative(self, start_node: int) -> Tuple[Dict[int, Optional[int]], List[int]]:
        """
        Iterative implementation of DFS using an explicit stack.

        This method avoids deep recursion, which can cause a stack overflow on very
        large or deep graphs. It manually manages the nodes to visit using a stack.
        
        :param start_node: The node to start the search from.
        :return: A tuple containing the parent map and the traversal path.
        """
        parent_map: Dict[int, Optional[int]] = {node: None for node in self.nodes}
        visited: Set[int] = set()
        traversal_path = []
        
        if start_node not in self.nodes:
            return {}, []
            
        # The stack for DFS, initialized with the start node.
        stack: List[int] = [start_node]

        while stack:
            # Pop a vertex from the top of the stack.
            u = stack.pop()
            
            # If the node has already been visited (e.g., added to stack multiple times
            # before being processed), skip it.
            if u not in visited:
                visited.add(u)
                traversal_path.append(u)

                # Push unvisited neighbors onto the stack.
                # We iterate through neighbors in reverse order to ensure that the
                # first neighbor (e.g., adj[0][0]) is processed first,
                # mimicking the typical recursive behavior.
                for v in reversed(self.graph.get(u, [])):
                    if v not in visited:
                        parent_map[v] = u # Parent is set when we decide to visit it
                        stack.append(v)

        return parent_map, traversal_path

# =================================================================================
# MAIN EXECUTION BLOCK - DEMONSTRATION FOR PROBLEM 11 (SIMPLE GRAPH)
# =================================================================================
if __name__ == '__main__':
    # --- 1. Graph Creation (Simple Graph from Val21, Fig 1.1) ---
    simple_graph = {
        0: [1, 3],
        1: [0, 4, 3],
        2: [0, 3, 5],
        3: [0, 1, 2, 4, 5, 6],
        4: [1, 3, 6],
        5: [2, 3, 6],
        6: [3, 4, 5]
    }
    node_names = {i: f"v{i+1}" for i in range(7)}
    start_node = 0

    # --- 2. Execution & Reporting ---
    solver = DfsSolverPython(simple_graph)
    
    print("=" * 60)
    print("DFS Algorithm Report for Simple Graph (Problem 11)")
    print(f"Source Node: '{node_names[start_node]}'")
    print("=" * 60)
    
    # --- Run and Print Recursive Version ---
    print("\n--- Recursive DFS ---")
    parents_rec, path_rec = solver.solve_recursive(start_node)
    path_str_rec = " -> ".join(node_names[n] for n in path_rec)
    print(f"  Traversal Path: {path_str_rec}")
    parent_of_v7_rec = parents_rec.get(6)
    print(f"  Parent of v7 ({node_names[6]}): {node_names[parent_of_v7_rec] if parent_of_v7_rec is not None else 'N/A'}")

    # --- Run and Print Iterative Version ---
    print("\n--- Iterative DFS ---")
    parents_iter, path_iter = solver.solve_iterative(start_node)
    path_str_iter = " -> ".join(node_names[n] for n in path_iter)
    print(f"  Traversal Path: {path_str_iter}")
    parent_of_v7_iter = parents_iter.get(6)
    print(f"  Parent of v7 ({node_names[6]}): {node_names[parent_of_v7_iter] if parent_of_v7_iter is not None else 'N/A'}")
    
    print("=" * 60)

File: combinatorics/resource/TMD/graph_tree/Code/problem11/problem11.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <map>
#include <string>
#include <algorithm>

// run with c++ 23
// A type alias for graph representation for better readability.
// We use an adjacency list for a simple graph.
using Graph = std::vector<std::vector<int>>;

/**
 * @class DfsSolver
 * @brief Encapsulates the logic and results of a Depth-First Search.
 *
 * This class provides both recursive and iterative implementations of DFS.
 * It stores results like parent pointers and the traversal path, which are
 * reset for each new run.
 */
class DfsSolver {
public:
    int num_nodes;
    std::vector<int> parent;         // Stores the predecessor in the DFS tree.
    std::vector<bool> visited;       // Tracks visited nodes.
    std::vector<int> traversal_path; // Records the order of visiting nodes.

    /**
     * @brief Constructs the DfsSolver.
     * @param n The total number of nodes in the graph.
     */
    DfsSolver(int n) : num_nodes(n) {}

    /**
     * @brief Public method to start the recursive DFS.
     *
     * This function initializes the necessary data structures and kicks off
     * the recursive exploration process from the start_node.
     *
     * @param adj The graph as an adjacency list.
     * @param start_node The node to start from.
     */
    void solve_recursive(const Graph& adj, int start_node) {
        // Initialization for a new run
        parent.assign(num_nodes, -1);
        visited.assign(num_nodes, false);
        traversal_path.clear();

        if (start_node >= num_nodes) {
            std::cerr << "Error: Start node is out of bounds." << std::endl;
            return;
        }
        
        // Start the recursive traversal
        dfs_util(adj, start_node);
    }
    
    /**
     * @brief Public method for the iterative DFS.
     *
     * This function uses an explicit stack to mimic the behavior of the
     * recursive call stack, exploring the graph in a depth-first manner.
     *
     * @param adj The graph as an adjacency list.
     * @param start_node The node to start from.
     */
    void solve_iterative(const Graph& adj, int start_node) {
        // Initialization for a new run
        parent.assign(num_nodes, -1);
        // `visited` is not a member here to show a different style
        std::vector<bool> local_visited(num_nodes, false); 
        traversal_path.clear();

        if (start_node >= num_nodes) {
            std::cerr << "Error: Start node is out of bounds." << std::endl;
            return;
        }
        
        // The explicit stack for the iterative DFS.
        std::stack<int> s;
        s.push(start_node);

        while (!s.empty()) {
            int u = s.top();
            s.pop();

            // Only process nodes that haven't been visited yet.
            if (!local_visited[u]) {
                local_visited[u] = true;
                traversal_path.push_back(u);

                // Push unvisited neighbors onto the stack.
                // We iterate in reverse to ensure that the first neighbor in the
                // adjacency list is the first one to be popped and visited.
                for (auto it = adj[u].rbegin(); it != adj[u].rend(); ++it) {
                    int v = *it;
                    if (!local_visited[v]) {
                        parent[v] = u; // Set parent before pushing
                        s.push(v);
                    }
                }
            }
        }
    }

private:
    /**
     * @brief Recursive utility function for DFS. This is the core of the recursive approach.
     *
     * This function embodies the recursive definition of DFS: visit a node, then
     * recursively visit each of its unvisited neighbors. The system's call stack
     * implicitly manages the "backtracking".
     *
     * @param adj The graph.
     * @param u The current node to visit.
     */
    void dfs_util(const Graph& adj, int u) {
        // 1. Base Action: Mark current node as visited and process it.
        visited[u] = true;
        traversal_path.push_back(u);

        // 2. Recursive Step: Recur for all unvisited adjacent vertices.
        for (int v : adj[u]) {
            if (!visited[v]) {
                parent[v] = u;
                dfs_util(adj, v);
            }
        }
        // Base Case: When this loop finishes, the function returns (backtracks).
    }
};

// =================================================================================
// MAIN FUNCTION - DEMONSTRATION FOR PROBLEM 11 (SIMPLE GRAPH)
// =================================================================================
int main() {
    int num_nodes = 7;
    int start_node = 0; // v1

    std::map<int, std::string> node_names;
    for(int i = 0; i < num_nodes; ++i) {
        node_names[i] = "v" + std::to_string(i + 1);
    }
    
    // --- Graph Creation (Simple Graph from Val21, Fig 1.1) ---
    Graph adj(num_nodes);
    adj[0] = {1, 3};
    adj[1] = {0, 4, 3};
    adj[2] = {0, 3, 5};
    adj[3] = {0, 1, 2, 4, 5, 6};
    adj[4] = {1, 3, 6};
    adj[5] = {2, 3, 6};
    adj[6] = {3, 4, 5};

    // --- Execution and Output ---
    DfsSolver solver(num_nodes);

    std::cout << "=====================================================" << std::endl;
    std::cout << "DFS Algorithm Report for Simple Graph (Problem 11)" << std::endl;
    std::cout << "Source Node: '" << node_names[start_node] << "'" << std::endl;
    std::cout << "=====================================================" << std::endl;
    
    // --- Run and Print Recursive Version ---
    solver.solve_recursive(adj, start_node);
    std::cout << "\n--- Recursive DFS ---" << std::endl;
    std::cout << "  Traversal Path: ";
    for(size_t i = 0; i < solver.traversal_path.size(); ++i) {
        std::cout << node_names[solver.traversal_path[i]] << (i == solver.traversal_path.size() - 1 ? "" : " -> ");
    }
    std::cout << "\n  Parent of v7: " 
              << (solver.parent[6] != -1 ? node_names[solver.parent[6]] : "N/A") 
              << std::endl;

    // --- Run and Print Iterative Version ---
    solver.solve_iterative(adj, start_node);
    std::cout << "\n--- Iterative DFS ---" << std::endl;
    std::cout << "  Traversal Path: ";
    for(size_t i = 0; i < solver.traversal_path.size(); ++i) {
        std::cout << node_names[solver.traversal_path[i]] << (i == solver.traversal_path.size() - 1 ? "" : " -> ");
    }
    std::cout << "\n  Parent of v7: " 
              << (solver.parent[6] != -1 ? node_names[solver.parent[6]] : "N/A") 
              << std::endl;

    std::cout << "\n=====================================================" << std::endl;

    return 0;
}